File: lintcode/naive/binary_tree_maximum_node/cpp/binary_tree_maximum_node2.cpp
#include <cmath>
#include <stack>

class Solution {
public:
    /**
     * @param root the root of binary tree
     * @return the max node
     */
    TreeNode* maxNode(TreeNode* root) 
    {
        if (root == nullptr) {
            return nullptr;
        }
        
        auto stack  = std::stack<TreeNode*>{ };
        auto curmax = root;
        
        stack.push(root);
        
        while (stack.size() > 0) 
        {
            const auto node = stack.top();
            
            curmax = std::max(node, curmax,
                [] (const TreeNode *left, const TreeNode *right) {
                    return left->val < right->val;
                });
            
            stack.pop();
            
            if (node->left != nullptr) {
                stack.push(node->left);
            }
            
            if (node->right != nullptr) {
                stack.push(node->right);
            }
        }
        
        return curmax;
    }
};

File: lintcode/naive/binary_tree_maximum_node/py/binary_tree_maximum_node2.py
class Solution:
    # @param {TreeNode} root the root of binary tree
    # @return {TreeNode} the max node
    def maxNode(self, root):
        if root == None:
            return None
            
        stack  = [root]
        curmax = root
        
        while len(stack) > 0:
            node   = stack.pop()
            curmax = max([node, curmax], key=lambda x: x.val)
            
            if node.left != None:
                stack.append(node.left)
                
            if node.right != None:
                stack.append(node.right)
                
        return curmax
